amplify/backend/function/fetchDomainBalance/src/services/networkBalance.js

```diff
@@ -0,0 +1,233 @@
+const { getColonyNetworkClient, Network } = require('@colony/colony-js');
+const {
+  providers,
+  constants: { AddressZero },
+} = require('ethers');
+const EnvVarsConfig = require('../config/envVars');
+const { getStartOfDayFor } = require('../utils');
+const { DEFAULT_TOKEN_DECIMALS } = require('../consts');
+const { getDomains, getAllColonyTokens } = require('../api/graphql/operations');
+const ExchangeRatesService = require('./exchangeRates');
+const { getTotalFiatAmountFor } = require('./tokens');
+const { getTokensDatesMap } = require('./actions');
+const NetworkConfig = require('../config/networkConfig');
+
+// Helper function to fetch block timestamp in milliseconds
+const getBlockTimestamp = async (provider, blockNumber) => {
+  const block = await provider.getBlock(blockNumber);
+  return block ? block.timestamp * 1000 : null; // convert to milliseconds
+};
+
+// Final check to compare left and right bounds for closest timestamp
+const checkFinalBounds = async ({
+  provider,
+  left,
+  right,
+  targetTimestamp,
+  closestBlock,
+  minTimestampDiff,
+}) => {
+  const leftTimestamp = await getBlockTimestamp(provider, left);
+  const rightTimestamp = await getBlockTimestamp(provider, right);
+
+  if (
+    leftTimestamp !== null &&
+    Math.abs(leftTimestamp - targetTimestamp) < minTimestampDiff
+  ) {
+    closestBlock = left;
+  }
+  if (
+    rightTimestamp !== null &&
+    Math.abs(rightTimestamp - targetTimestamp) < minTimestampDiff
+  ) {
+    closestBlock = right;
+  }
+
+  return closestBlock;
+};
+
+// Finds the closest past block to a target date using binary search
+const findClosestPastBlock = async ({
+  targetDate,
+  averageBlockTime,
+  currentBlockNumber,
+  provider,
+}) => {
+  if (!targetDate) {
+    return currentBlockNumber;
+  }
+  const targetTimestamp = new Date(targetDate).getTime();
+
+  // Step 1: Estimate initial block guess near target timestamp
+  const currentBlockTimestamp = await getBlockTimestamp(
+    provider,
+    currentBlockNumber,
+  );
+  const timeDiffSeconds = (currentBlockTimestamp - targetTimestamp) / 1000;
+  let estimatedBlock =
+    currentBlockNumber - Math.floor(timeDiffSeconds / averageBlockTime);
+
+  // Step 2: Set up binary search bounds
```

Comment:
This is a nice solution, but is it actually worth the extra computational effort? Your initial estimation might be a few seconds off, which is unlikely to matter when fetching past balance (it would only matter if the balance changed exactly within that few-second window). On the other hand, doing binary search will slow down the lambda at each execution.

Consider fetching a balance from 30 days ago. The estimated block number will be approximately n - 10,368,000 (where n is the current block number), so your initial bounds will be [n - 10,369,000, n]. The target block is much closer to the left bound than to the right one, and in the worst case scenario it will take roughly 24 calls (log₂ of the search range) to `getBlock()` to find the most precise block.

Unless I'm missing something else, this seems like a high price to pay for getting a block that is only a few seconds "more correct", if that makes sense.

Reply:
As I mentioned in our chat, your comment is very valid and worth considering, so I'll proceed creating another issue for this improvement 🙌 thanks so much for your review!!